### PR TITLE
CObject fixes

### PR DESCRIPTION
**Pull request by waliedyassen**
Fixes 1 bug with `CObject::DeleteAllTempObjectsInArea`, implements 3 missing behaviors for `MI_RCBOMB`, `MI_BEACHBALL`  and `MI_PETROLPUMP2` and all the missing damage effects.

This should fix `The Shootist` mission problem in  #782.

### CODE REVIEW

File: src/core/General.h [hunk 1]
```diff
@@ -159,6 +159,11 @@ class CGeneral
 		
 	static int32 GetRandomNumberInRange(int32 low, int32 high)
 		{ return low + (high - low)*(GetRandomNumber()/float(MYRAND_MAX + 1)); }
+
+	// Returns inclusive value in the specified range
+	static int32 GetRandomNumberInRangeInc(int32 low, int32 high)
+		{ return GetRandomNumberInRange(low - 1, high + 1); }
+
```
Comment (Sergeanur) [inline src/core/General.h:166]: We don't need this.

File: src/objects/Object.cpp [hunk 2]
```diff
@@ -107,14 +109,65 @@ CObject::ProcessControl(void)
 		m_vecMoveSpeed *= fTimeStep;
 		m_vecTurnSpeed *= fTimeStep;
 	}
-	if ((GetModelIndex() == MI_EXPLODINGBARREL || GetModelIndex() == MI_PETROLPUMP) && bHasBeenDamaged && bIsVisible
+	auto mi = GetModelIndex();
```
Comment (Sergeanur) [inline src/objects/Object.cpp:112]: We don't use auto.

File: src/objects/Object.cpp [hunk 3]
```diff
@@ -107,14 +109,65 @@ CObject::ProcessControl(void)
...
+	if (mi == MI_RCBOMB) {
+		float fTurnForce = -(m_fTurnMass / 20.0f);
+		CPhysical::ApplyTurnForce(m_vecMoveSpeed * fTurnForce, -GetForward());
+		float fScalar = 1.0f - m_vecMoveSpeed.MagnitudeSqr() / 5.0f;
+		float fScalarTimed = Pow(fScalar, CTimer::GetTimeStep());
+		m_vecMoveSpeed *= fScalarTimed;
+	}
+	if (mi == MI_BEACHBALL) {
+		constexpr uint8 BEACHBALL_MAX_SCORE = 250;
+		constexpr float BEACHBALL_DEACCELERATION = 2.5f;
```
Comment (Sergeanur) [inline src/objects/Object.cpp:130]: We don't use constexpr. Use #define instead.

File: src/objects/Object.cpp [hunk 4]
```diff
@@ -107,14 +109,65 @@ CObject::ProcessControl(void)
...
+		float fMoveSpeedMag = m_vecMoveSpeed.Magnitude2D();
+		float fTimeScale = powf(0.95, CTimer::GetTimeStep());
```
Comment (Sergeanur) [inline src/objects/Object.cpp:132]: `Pow`. Also in here and other occasions below floating point value needs `f` suffix

File: src/objects/Object.cpp [hunk 5]
```diff
@@ -107,14 +109,65 @@ CObject::ProcessControl(void)
...
+			float distanceMagnitude = distance.Magnitude2D();
+			if (distance.z > 0.0 && distance.z < 1.5f && distanceMagnitude < 1.0) {
+				CVector playerSpeed = FindPlayerSpeed();
+				if (fMoveSpeedMag < 0.05 && playerSpeed.Magnitude() > 0.1) {
+					playerSpeed.z = 0.0f;
+					playerSpeed.Normalise();
+					playerSpeed.z = 0.3;
+					m_vecMoveSpeed = CVector(playerSpeed.x / BEACHBALL_DEACCELERATION, playerSpeed.y / BEACHBALL_DEACCELERATION, 1.0f / BEACHBALL_DEACCELERATION * 0.3);
+					PlayOneShotScriptObject(SCRIPT_SOUND_HIT_BALL, GetPosition());
+					if (m_nBeachballBounces > 0) {
```
Comment (Sergeanur) [inline src/objects/Object.cpp:147]: Missing turn speed calculation above this

File: src/objects/Object.cpp [hunk 6]
```diff
@@ -107,14 +109,65 @@ CObject::ProcessControl(void)
...
+						m_nBeachballBounces++;
+						sprintf(gString, "%d", m_nBeachballBounces);
```
Comment (Sergeanur) [inline src/objects/Object.cpp:149]: Missing another `if (m_nBeachballBounces > 0) {` above this.

File: src/objects/Object.cpp [hunk 7]
```diff
@@ -107,14 +109,65 @@ CObject::ProcessControl(void)
...
+		constexpr uint8 BEACHBALL_MAX_SCORE = 250;
+		constexpr float BEACHBALL_DEACCELERATION = 2.5f;
+		float fMoveSpeedMag = m_vecMoveSpeed.Magnitude2D();
+		float fTimeScale = powf(0.95, CTimer::GetTimeStep());
+		m_vecMoveSpeed.x *= fTimeScale;
+		m_vecMoveSpeed.y *= fTimeScale;
+		m_vecMoveSpeed.z += fMoveSpeedMag - m_vecMoveSpeed.Magnitude2D();
+		if (!FindPlayerVehicle()) {
+			CVector distance = FindPlayerCoors() - GetPosition();
+			float distanceMagnitude = distance.Magnitude2D();
```
Comment (Sergeanur) [inline src/objects/Object.cpp:138]: A bit of a nit, but original code didn't use variable in here and instead called `distance.Magnitude2D()` both times.

File: src/objects/Object.cpp [hunk 8]
```diff
@@ -107,14 +109,65 @@ CObject::ProcessControl(void)
...
+						m_nBeachballBounces++;
+						sprintf(gString, "%d", m_nBeachballBounces);
+						CMoneyMessages::RegisterOne(GetPosition(), gString, 255, 50, 0, 0.6f, 0.5f);
+						CStats::RegisterHighestScore(3, m_nBeachballBounces);
+					}
+				}
+			}
+			if (distance.z > -1.05 && distance.z < -0.6 && distanceMagnitude < 0.9 && m_vecMoveSpeed.z < 0.0f) {
+				m_vecMoveSpeed.x += CGeneral::GetRandomNumberInRangeInc(-3, 4) / 100.0;
+				m_vecMoveSpeed.y += CGeneral::GetRandomNumberInRangeInc(-3, 4) / 100.0;
```
Comment (Sergeanur) [inline src/objects/Object.cpp:157]:
```suggestion
				m_vecMoveSpeed.x += (CGeneral::GetRandomNumber() % 8 - 3) / 100.0f;
				m_vecMoveSpeed.y += (CGeneral::GetRandomNumber() % 8 - 3) / 100.0f;
```

File: src/objects/Object.cpp [hunk 9]
```diff
@@ -107,14 +109,65 @@ CObject::ProcessControl(void)
...
+				m_vecMoveSpeed.z = Max(m_vecMoveSpeed.z + 0.3f, 0.2f);
+				PlayOneShotScriptObject(SCRIPT_SOUND_HIT_BALL, GetPosition());
+				m_vecTurnSpeed.x += CGeneral::GetRandomNumberInRangeInc(-7, 8) / 10.0f;
+				m_vecTurnSpeed.y += CGeneral::GetRandomNumberInRangeInc(-7, 8) / 10.0f;
```
Comment (Sergeanur) [inline src/objects/Object.cpp:161]:
```suggestion
				m_vecTurnSpeed.x += (CGeneral::GetRandomNumber() % 16 - 7) / 10.0f;
				m_vecTurnSpeed.y += (CGeneral::GetRandomNumber() % 16 - 7) / 10.0f;
```

File: src/objects/Object.cpp [hunk 10]
```diff
@@ -107,14 +109,65 @@ CObject::ProcessControl(void)
...
+				if (++m_nBeachballBounces >= BEACHBALL_MAX_SCORE) {
+					m_nBeachballBounces = BEACHBALL_MAX_SCORE;
+				}
```
Comment (Sergeanur) [inline src/objects/Object.cpp:164]:
```suggestion
				m_nBeachballBounces++;
				m_nBeachballBounces = Max(m_nBeachballBounces, BEACHBALL_MAX_SCORE);
```

File: src/objects/Object.cpp [hunk 11]
```diff
@@ -107,14 +109,65 @@ CObject::ProcessControl(void)
...
+				sprintf(gString, "%d", m_nBeachballBounces);
+				CMoneyMessages::RegisterOne(GetPosition(), gString, 255, 50, 0, 0.6f, 0.5f);
+				CStats::RegisterHighestScore(3, m_nBeachballBounces);
+			}
+		}
+	}
```
Comment (Sergeanur) [inline src/objects/Object.cpp:170]: Some flag code is missing in the end of function.

**Review by Sergeanur — CHANGES_REQUESTED** [review]
Thank you for your contribution, there's a couple of things to be fixed though.

File: src/objects/Object.cpp (hunk 3 above)
Comment (ShFil119) [reply src/objects/Object.cpp:130]: or just `static const`

Comment (Sergeanur) [reply src/objects/Object.cpp:130]: But not in this occasion, please.

File: src/objects/Object.cpp (hunk 10 above)
Comment (waliedyassen) [reply src/objects/Object.cpp:164]: Max wouldn't work there, I assume you meant Min instead

Comment (Sergeanur) [reply src/objects/Object.cpp]: Indeed

File: src/objects/Object.cpp [hunk 12]
```diff
@@ -287,132 +349,320 @@ CObject::ObjectDamage(float amount)
...
+					CParticle::AddParticle(PARTICLE_CAR_DEBRIS, vecPos, vecDir, nil, fSize, selectedColor, nRotationSpeed, 0, nCurFrame, 0);
+				}
+				if (m_nCollisionDamageEffect == DAMAGE_EFFECT_SMASH_BLACKBAG) {
+					PlayOneShotScriptObject(SCRIPT_SOUND_BOX_DESTROYED_2, vecPos);
+				} else if (m_nCollisionDamageEffect == DAMAGE_EFFECT_SMASH_BEACHLOUNGE_WOOD) {
+					PlayOneShotScriptObject(SCRIPT_SOUND_METAL_COLLISION, vecPos);
+				}
+				break;
+			}
+			default:
```
Comment (aap) [inline src/objects/Object.cpp:663]: where is DAMAGE_EFFECT_SMASH_VEGPALM?

File: src/objects/Object.cpp [hunk 13]
```diff
@@ -107,19 +112,76 @@ CObject::ProcessControl(void)
...
+	if (mi == MI_BEACHBALL) {
+		float fTimeStep = Pow(0.95f, CTimer::GetTimeStep());
+		float fPreviousVecSpeedMag = m_vecMoveSpeed.Magnitude2D();
+		m_vecMoveSpeed.x *= fTimeStep;
+		m_vecMoveSpeed.y *= fTimeStep;
+		m_vecMoveSpeed.z += fPreviousVecSpeedMag - m_vecMoveSpeed.Magnitude2D();
+		if (!FindPlayerVehicle()) {
+			CVector distance = FindPlayerCoors() - GetPosition();
+			float distanceMagnitude = distance.Magnitude2D();
+			if (distance.z > 0.0 && distance.z < 1.5f && distanceMagnitude < 1.0) {
```
[the remark this answers is not in the document]
Comment (waliedyassen) [reply src/objects/Object.cpp:140]: It's used in the other condition, in the mobile binary they have it as a variable like this.

[the remark this answers is not in the document]
Comment (aap) [reply src/objects/Object.cpp:140]: That's a compiler optimization. the android compiler was much more aggressive

[the remark this answers is not in the document]
Comment (waliedyassen) [reply src/objects/Object.cpp:140]: Makes sense

File: src/objects/Object.cpp [hunk 14]
```diff
@@ -287,132 +349,320 @@ CObject::ObjectDamage(float amount)
...
+			case DAMAGE_EFFECT_SMASH_COMPLETELY:
+			case DAMAGE_EFFECT_CHANGE_THEN_SMASH:
+			{
+				if (m_nCollisionDamageEffect == DAMAGE_EFFECT_CHANGE_THEN_SMASH && !bRenderDamaged) {
+					bRenderDamaged = true;
+					return;
+				}
 				bIsVisible = false;
 				bUsesCollision = false;
+				if (!GetIsStatic()) {
```
[the remark this answers is not in the document]
Comment (waliedyassen) [reply src/objects/Object.cpp:403]: Are you sure about that one? Looks like it is the way it is right now for me.

File: src/objects/Object.cpp [hunk 15]
```diff
@@ -287,132 +349,320 @@ CObject::ObjectDamage(float amount)
...
+			case DAMAGE_EFFECT_SPLIT_MODEL:
+				return;
+			case DAMAGE_EFFECT_SMASH_AND_DAMAGE_TRAFFICLIGHTS:
+			{
+				static RwRGBA debrisColor = { 0xc8,0xc8,0xc8,0xff };
+				if (bRenderDamaged) {
+					break;
+				}
 				bRenderDamaged = true;
+				CBaseModelInfo *modelInfo = CModelInfo::GetModelInfo(GetModelIndex());
```
[the remark this answers is not in the document]
Comment (waliedyassen) [reply src/objects/Object.cpp:367]: Will do, but that would not match the binary anymore, to be more accurate, in the binary it's something like this:
```cpp
CVector min = 0.85f * CModelInfo::GetModelInfo(GetModelIndex())->GetColModel()->boundingBox.min;
CVector max = 0.85f * CModelInfo::GetModelInfo(GetModelIndex())->GetColModel()->boundingBox.max;
```

File: src/objects/Object.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (aap) [reply src/objects/Object.cpp:403]: oh yeah, i got confused. logically this makes no sense to me. why would you remove non-static objects from the moving list???

File: src/objects/Object.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (aap) [reply src/objects/Object.cpp:367]: you don't have to go through modelinfo manually, GetColModel takes care of that. but yes, it looks like they didn't even use a temporary for the col model, so just
```
CVector min = 0.85f * GetColModel()->boundingBox.min;
CVector max = 0.85f * GetColModel()->boundingBox.max;
`

File: src/objects/Object.cpp [hunk 16]
```diff
@@ -14,15 +14,20 @@
...
 
 int16 CObject::nNoTempObjects;
 //int16 CObject::nBodyCastHealth = 1000;
 float CObject::fDistToNearestTree;
 
-void *CObject::operator new(size_t sz) { return CPools::GetObjectPool()->New();  }
-void *CObject::operator new(size_t sz, int handle) { return CPools::GetObjectPool()->New(handle);};
-void CObject::operator delete(void *p, size_t sz) { CPools::GetObjectPool()->Delete((CObject*)p); }
-void CObject::operator delete(void *p, int handle) { CPools::GetObjectPool()->Delete((CObject*)p); }
+void *CObject::operator new(size_t sz) { return CPools::GetObjectPool()->New(); }
```
[the remark this answers is not in the document]
Comment (waliedyassen) [reply src/objects/Object.cpp:27]: This was a revert to the style changes I did in a previous commit, I'll go through and make sure it is matching.

File: src/core/General.h [hunk 17]
```diff
@@ -159,6 +159,7 @@ class CGeneral
 		
 	static int32 GetRandomNumberInRange(int32 low, int32 high)
 		{ return low + (high - low)*(GetRandomNumber()/float(MYRAND_MAX + 1)); }
+
```
Comment (Sergeanur) [inline src/core/General.h:162]: Please remove this empty line change.

Comment (waliedyassen) [reply src/core/General.h]: Done

Comment (Sergeanur) [reply src/core/General.h]: Eh, now there's a change at the end of it (

Comment (waliedyassen) [reply src/core/General.h]: That matches the miami branch, the line was an extra line as well

Comment (Sergeanur) [reply src/core/General.h]: It wouldn't show it as a change if it's a match.

Comment (waliedyassen) [reply src/core/General.h]: Sorry, you are right. fixed it. But that's weird, it only shows when checking out locally. The empty lines are removed when viewing in the GitHub code viewer ([like here](https://github.com/GTAmodding/re3/blob/miami/src/core/General.h))

**Review by Sergeanur — APPROVED** [review]

**Pull request merged**